src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,445 @@
+/// with a dot
+/// \param i: an integer expression
+/// \param max_size: a maximal size for the string
+/// \param ref_type: a type for refined strings
+/// \return a string expression
+string_exprt string_constraint_generatort::add_axioms_for_fractional_part(
+    const exprt &i, size_t max_size, const refined_string_typet &ref_type)
+{
+  string_exprt res=fresh_string(ref_type);
+  const typet &type=i.type();
+  assert(type.id()==ID_signedbv);
+  exprt ten=from_integer(10, type);
+  const typet &char_type=ref_type.get_char_type();
+  const typet &index_type=ref_type.get_index_type();
+  exprt zero_char=constant_char('0', char_type);
+  exprt nine_char=constant_char('9', char_type);
+  exprt max=from_integer(max_size, index_type);
+
+  // We add axioms:
+  // a1 : 2 <= |res| <= max_size
+  // a2 : forall 1 <= i < size '0' < res[i] < '9'
+  // res[0] = '.'
+  // a3 : i = sum_j 10^j res[j] - '0'
+  // for all j : !(|res| = j+1 && res[j]='0')
+  // for all j : |res| <= j => res[j]='0'
+
+  and_exprt a1(res.axiom_for_is_strictly_longer_than(1),
+               res.axiom_for_is_shorter_than(max));
+  axioms.push_back(a1);
+
+  equal_exprt starts_with_dot(res[0], from_integer('.', char_type));
+
+  exprt::operandst digit_constraints;
+  digit_constraints.push_back(starts_with_dot);
+  exprt sum=from_integer(0, type);
+
+  for(size_t j=1; j<max_size; j++)
+  {
+    binary_relation_exprt after_end(
+      res.length(), ID_le, from_integer(j, res.length().type()));
+    mult_exprt ten_sum(sum, ten);
+
+    // sum = 10 * sum + (res[j]-'0')
+    if_exprt to_add(
+      after_end,
+      from_integer(0, type),
+      typecast_exprt(minus_exprt(res[j], zero_char), type));
+    sum=plus_exprt(ten_sum, to_add);
+
+    and_exprt is_number(
+          binary_relation_exprt(res[j], ID_ge, zero_char),
+          binary_relation_exprt(res[j], ID_le, nine_char));
+    digit_constraints.push_back(is_number);
+
+    // There are no trailing zeros except for ".0" (i.e j=2)
+    if(j>1)
+    {
+      not_exprt no_trailing_zero(and_exprt(
+        equal_exprt(res.length(), from_integer(j+1, res.length().type())),
+      equal_exprt(res[j], zero_char)));
+      axioms.push_back(no_trailing_zero);
+    }
+  }
+
+  exprt a2=conjunction(digit_constraints);
+  axioms.push_back(a2);
+
+  equal_exprt a3(i, sum);
+  axioms.push_back(a3);
+
+  return res;
+}
+
+/// Add axioms to write the float in scientific notation.
+///
+/// A float is represented as $f = m * 2^e$ where $0 <= m < 2$ is the
+/// significand and $-126 <= e <= 127$ is the exponent.
+/// We want an alternate representation by finding $n$ and
+/// $d$ such that $f=n*10^d$. We can estimate $d$ the following way:
+/// $d ~= log_10(f/n) ~= log_10(m) + log_10(2) * e - log_10(n)$
+/// $d = floor(log_10(2) * e)$
+/// Then $n$ can be expressed by the equation:
+/// $log_10(n) = log_10(m) + log_10(2) * e - d$
+/// $n = f /10^d = m * 2^e / 10^d = m * 2^e / 10^(floor(log_10(2) * e))$
```

Comment:
No, once `d` is fixed this is exact, what can happen is that the estimated `d` is to small and in that case we have add 1 to `d` and divide `n` by 10 to get the actual representation.